### DISCUSSION TITLE
Data type objects for SimpleArray

### DISCUSSION THREAD

**Discussion by tigercosmos** [solvcon/modmesh#427] [Ideas]
Currently, SimpleArray only supports dtype as numbers. It is possible for SimpleArray to support dtype as an object. Reference: https://numpy.org/doc/stable/reference/arrays.dtypes.html

**Comment by yungyuc** [comment]
We should start with accepting dtype as a string before SimpleArray is enhanced to support struct array.  String suffices for fundamental types.

**Reply by tigercosmos** [reply]
You meant like this `arr_str = np.array(['hello', 'world'], dtype='S10')`?
Yes, I think it can be the first step. We also need to introduce a data type descriptor. A data type enum is not enough anymore.

**Reply by yungyuc** [reply]
What an enum cannot do?

**Reply by tigercosmos** [reply]
We know `int` is 4 bytes, `double` is 8 bytes, so we can get the "bytes length" from the enum. However, for an arbitrary string type, e.g. "S10", "S13", we need to record the length information as well. 

For example, for "S13":
```
struct Descriptor {
    DataType type; // DataType::FixedString
    size_t size; // 13
    // ...
};
```

**Reply by yungyuc** [reply]
I am not seeing a use case for using SimpleArray to store strings.  You need dynamic allocation for each of the string element in the array.  It is inefficient.

Do you know a use case for efficient processing of string using SimpleArray?

**Reply by tigercosmos** [reply]
Actually I also wonder what is the use case to store objects or strings into ndarray? I want to support this because ndarray can do it.

**Reply by yungyuc** [reply]
String and object arrays in numpy are slow.  We do not need to support them in SimpleArray and modmesh because we will not use the string and object arrays in C++.

For the string specifiers for dtype, I propose to use the convention of `int8`, `uint64`, `float32`, etc.  It is like the fixed-width type names in C++ and is explicit about bit width for real numbers.

**Reply by tigercosmos** [reply]
I think we have already supported it except for the `bool` type. For example, the statement is valid as `modmesh.SimpleArray((3, 2), value=-2, dtype='int64')`. Is this what you meant?

**Reply by yungyuc** [reply]
Yeah.  That's what I meant.

**Reply by tigercosmos** [reply]
Then we don't need to do anything. Also, the original topic is rejected. Close this discussion.

**Reply by yungyuc** [reply]
The topic can remain open for discussion if you want
